File: Components/WinForms/VB/EditorsMainDemo/Modules/ModuleTokenEdit.vb
Imports System
Imports System.Drawing
Imports DevExpress.Utils
Imports DevExpress.XtraEditors.Repository

Namespace DevExpress.XtraEditors.Demos
	Partial Public Class ModuleTokenEdit
		Inherits TutorialControl
		Private properties As RepositoryItemTokenEdit
		Public Sub New()
            InitializeComponent()
            Me.properties = tokenEdit.Properties
            TutorialInfo.WhatsThisCodeFile = "VB\EditorsMainDemo\Modules\ModuleTokenEdit.vb"
            TutorialInfo.WhatsThisXMLFile = "tokenEdit.xml"
		End Sub
		Protected Overrides Sub OnLoad(ByVal e As EventArgs)
			MyBase.OnLoad(e)
			InitValues()
		End Sub
		Private Sub InitValues()
			Me.ceShowDropDown.Checked = Me.properties.ShowDropDown
			Me.seDropDownRowCount.Value = Me.properties.DropDownRowCount
            Me.cbeCheckMode.Properties.Items.AddEnum(GetType(TokenEditCheckMode))
            Me.cbeCheckMode.EditValue = Me.properties.CheckMode
            Me.seMaxExpandLines.Value = Me.properties.MaxExpandLines
            Me.ceShowPopupPanel.Checked = Me.properties.PopupPanelOptions.ShowPopupPanel
            Me.cbePopupPanelLocation.Properties.Items.AddEnum(GetType(TokenEditPopupPanelLocation))
            Me.cbePopupPanelLocation.EditValue = Me.properties.PopupPanelOptions.Location
            Me.cbeDropDownMode.Properties.Items.AddEnum(GetType(TokenEditDropDownShowMode))
            Me.cbeDropDownMode.EditValue = Me.properties.DropDownShowMode
            Me.cbeEditMode.Properties.Items.AddEnum(GetType(TokenEditMode))
            Me.cbeEditMode.EditValue = Me.properties.EditMode
            Me.cbeFontSize.Properties.Items.AddEnum(GetType(TokenEditFontSize))
            Me.cbeFontSize.EditValue = TokenEditFontSize.Default
            Me.ceShowItemRemoveButton.Checked = Me.properties.ShowRemoveTokenButtons
            Me.cePopupSizeable.Checked = Me.properties.PopupSizeable
            Me.cbeGlyphLocation.Properties.Items.AddEnum(GetType(TokenEditGlyphLocation))
            Me.cbeGlyphLocation.EditValue = Me.properties.TokenGlyphLocation
            If Me.cbeEditValueSep.Properties.Items.Count > 0 Then
                Me.cbeEditValueSep.SelectedIndex = 0
            End If
            Me.tokenEdit.EditValue = "January"
        End Sub
        ' <ceShowDropDown>
        Private Sub OnShowDropDownCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles ceShowDropDown.CheckedChanged
            Dim checkEdit As CheckEdit = CType(sender, CheckEdit)
            Me.properties.ShowDropDown = checkEdit.Checked
        End Sub
        ' </ceShowDropDown>
        ' <btnClearEventLog>
        Private Sub OnClearEventLogClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnClearEventLog.Click
            Me.eventsLog.Text = String.Empty
        End Sub
        ' </btnClearEventLog>
        ' <tokenEdit>
        Private Sub OnTokenEditBeforeShowPopupPanel(ByVal sender As Object, ByVal e As TokenEditBeforeShowPopupPanelEventArgs) Handles tokenEdit.Properties.BeforeShowPopupPanel
            AddEventEntry("BeforeShowPopupPanel", e.Description)
            Me.tokenLabelValue.Text = e.Description
        End Sub
        Private Sub OnTokenEditTokenMouseEnter(ByVal sender As Object, ByVal e As TokenEditTokenMouseEnterEventArgs) Handles tokenEdit.Properties.TokenMouseEnter
            AddEventEntry("TokenMouseEnter", e.Description)
        End Sub
        Private Sub OnTokenEditTokenMouseHover(ByVal sender As Object, ByVal e As TokenEditTokenMouseHoverEventArgs) Handles tokenEdit.Properties.TokenMouseHover
            AddEventEntry("TokenMouseHover", e.Description)
        End Sub
        Private Sub OnTokenEditTokenMouseLeave(ByVal sender As Object, ByVal e As TokenEditTokenMouseLeaveEventArgs) Handles tokenEdit.Properties.TokenMouseLeave
            AddEventEntry("TokenMouseLeave", e.Description)
        End Sub
        Private Sub OnTokenEditValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles tokenEdit.EditValueChanged
            If Me.properties Is Nothing Then
                Return
            End If
            Dim tokenEdit As TokenEdit = CType(sender, TokenEdit)
            AddEventEntry("EditValueChanged", tokenEdit.EditValue)
            UpdateEditValue()
        End Sub
        Private Sub OnTokenEditTokenClick(ByVal sender As Object, ByVal e As TokenEditTokenClickEventArgs) Handles tokenEdit.Properties.TokenClick
            Dim tokenEdit As TokenEdit = CType(sender, TokenEdit)
            AddEventEntry("TokenClick", e.Description)
        End Sub
        Private Sub OnTokenEditTokenDoubleClick(ByVal sender As Object, ByVal e As TokenEditTokenClickEventArgs) Handles tokenEdit.Properties.TokenDoubleClick
            Dim tokenEdit As TokenEdit = CType(sender, TokenEdit)
            AddEventEntry("TokenDoubleClick", e.Description)
        End Sub
        Private Sub OnTokenEditCheckStateChanged(ByVal sender As Object, ByVal e As EventArgs) Handles tokenEdit.Properties.TokenCheckStateChanged
            Dim tokenEdit As TokenEdit = CType(sender, TokenEdit)
            AddEventEntry("TokenCheckStateChanged", Nothing)
        End Sub
        ' </tokenEdit>
        ' <seDropDownRowCount>
        Private Sub OnDropDownRowCountEditValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles seDropDownRowCount.EditValueChanged
            If Me.properties Is Nothing Then
                Return
            End If
            Dim spinEdit As SpinEdit = CType(sender, SpinEdit)
            Me.properties.DropDownRowCount = CInt(Fix(spinEdit.Value))
        End Sub
        ' </seDropDownRowCount>
        ' <cbeCheckMode>
        Private Sub OnCheckModeSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbeCheckMode.SelectedIndexChanged
            Dim comboBox As ComboBoxEdit = CType(sender, ComboBoxEdit)
            Me.properties.CheckMode = CType(comboBox.EditValue, TokenEditCheckMode)
        End Sub
        ' </cbeCheckMode>
        ' <seMaxExpandLines>
        Private Sub OnMaxExpandLinesEditValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles seMaxExpandLines.EditValueChanged
            If Me.properties Is Nothing Then
                Return
            End If
            Dim spinEdit As SpinEdit = CType(sender, SpinEdit)
            Me.properties.MaxExpandLines = CInt(Fix(spinEdit.Value))
        End Sub
        ' </seMaxExpandLines>
        ' <ceShowPopupPanel>
        Private Sub OnShowPopupPanelCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles ceShowPopupPanel.CheckedChanged
            Dim checkEdit As CheckEdit = CType(sender, CheckEdit)
            Me.properties.PopupPanelOptions.ShowPopupPanel = checkEdit.Checked
        End Sub
        ' </ceShowPopupPanel>
        ' <cbePopupPanelLocation>
        Private Sub OnPopupPanelLocationSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbePopupPanelLocation.SelectedIndexChanged
            Dim comboBox As ComboBoxEdit = CType(sender, ComboBoxEdit)
            Me.properties.PopupPanelOptions.Location = CType(comboBox.EditValue, TokenEditPopupPanelLocation)
        End Sub
        ' </cbePopupPanelLocation>
        ' <ceShowItemRemoveButton>
        Private Sub OnShowItemRemoveButtonCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles ceShowItemRemoveButton.CheckedChanged
            Dim checkEdit As CheckEdit = CType(sender, CheckEdit)
            Me.properties.ShowRemoveTokenButtons = checkEdit.Checked
        End Sub
        ' </ceShowItemRemoveButton>
        ' <cePopupSizeable>
        Private Sub OnPopupSizeableCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cePopupSizeable.CheckedChanged
            Dim checkEdit As CheckEdit = CType(sender, CheckEdit)
            Me.properties.PopupSizeable = checkEdit.Checked
        End Sub
        ' </cePopupSizeable>
        ' <cbeEditValueSep>
        Private Sub OnEditValueSepSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbeEditValueSep.SelectedIndexChanged
            Dim comboBox As ComboBoxEdit = CType(sender, ComboBoxEdit)
            Me.properties.EditValueSeparatorChar = Char.Parse(comboBox.EditValue.ToString())
        End Sub
        ' </cbeEditValueSep>
        ' <cbeDropDownMode>
        Private Sub OnDropDownModeSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbeDropDownMode.SelectedIndexChanged
            Dim comboBox As ComboBoxEdit = CType(sender, ComboBoxEdit)
            Me.properties.DropDownShowMode = CType(comboBox.EditValue, TokenEditDropDownShowMode)
        End Sub
        ' </cbeDropDownMode>
        Private Sub UpdateEditValue()
            Dim value As String = TryCast(tokenEdit.EditValue, String)
            Me.editValueLabel.Text = String.Format("Edit Value: {0}", If(String.IsNullOrEmpty(value), "(empty)", value))
        End Sub
        ' <cbeEditMode>
        Private Sub OnEditModeSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbeEditMode.SelectedIndexChanged
            Dim comboBox As ComboBoxEdit = CType(sender, ComboBoxEdit)
            Me.properties.EditMode = CType(comboBox.EditValue, TokenEditMode)
        End Sub
        ' </cbeEditMode>
        ' <cbeFontSize>
        Private Sub OnFontSizeSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbeFontSize.SelectedIndexChanged
            Dim defaultFont As Font = AppearanceObject.DefaultFont, newFont As Font = Nothing
            Dim comboBox As ComboBoxEdit = CType(sender, ComboBoxEdit)
            Dim fontSize As TokenEditFontSize = CType(comboBox.EditValue, TokenEditFontSize)
            Select Case fontSize
                Case TokenEditFontSize.Default
                    newFont = defaultFont
                Case TokenEditFontSize.Medium
                    newFont = New Font(defaultFont.FontFamily, defaultFont.SizeInPoints + 2.5F)
                Case TokenEditFontSize.Large
                    newFont = New Font(defaultFont.FontFamily, defaultFont.SizeInPoints + 3.75F)
            End Select
            Me.properties.AppearanceDropDown.Font = newFont
            Me.properties.Appearance.Font = Me.properties.AppearanceDropDown.Font
        End Sub
        ' </cbeFontSize>
        ' <cbeGlyphLocation>
        Private Sub OnGlyphAlignmentModeSelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbeGlyphLocation.SelectedIndexChanged
            Dim comboBox As ComboBoxEdit = CType(sender, ComboBoxEdit)
            Me.properties.TokenGlyphLocation = CType(comboBox.EditValue, TokenEditGlyphLocation)
        End Sub
        ' </cbeGlyphLocation>
        Private Sub AddEventEntry(ByVal entry As String, ByVal obj As Object)
            Dim val As String = String.Empty
            If obj IsNot Nothing Then
                val = If(TypeOf obj Is TokenEdit, GetType(TokenEdit).Name, obj.ToString())
            Else
                val = "(null)"
            End If
            Me.eventsLog.Text = String.Format("{0}: {1}{2}", entry, val, Environment.NewLine) + eventsLog.Text
        End Sub
    End Class

	Friend Enum TokenEditFontSize
		[Default]
		Medium
		Large
	End Enum
End Namespace

File: Components/WinForms/CS/EditorsMainDemo/Modules/ModuleTokenEdit.cs
using System;
using System.Drawing;
using DevExpress.Utils;
using DevExpress.XtraEditors.Repository;

namespace DevExpress.XtraEditors.Demos {
    public partial class ModuleTokenEdit : TutorialControl {
        RepositoryItemTokenEdit properties;
        public ModuleTokenEdit() {
            InitializeComponent();
            this.properties = tokenEdit.Properties;
            TutorialInfo.WhatsThisCodeFile = "CS\\EditorsMainDemo\\Modules\\ModuleTokenEdit.cs";
            TutorialInfo.WhatsThisXMLFile = "DevExpress.XtraEditors.Demos.CodeInfo.tokenEdit.xml";
        }
        protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            InitValues();
        }
        protected override void UpdateTabControlPosition(XtraTab.XtraTabControl tabControl) { }
        protected override void CalcContentCore() {
            CalcLayoutControlBounds(layoutControl1);
        }
        void InitValues() {
            this.ceShowDropDown.Checked = this.properties.ShowDropDown;
            this.seDropDownRowCount.Value = this.properties.DropDownRowCount;
            this.cbeCheckMode.Properties.Items.AddEnum(typeof(TokenEditCheckMode));
            this.cbeCheckMode.EditValue = this.properties.CheckMode;
            this.seMaxExpandLines.Value = this.properties.MaxExpandLines;
            this.ceShowPopupPanel.Checked = this.properties.PopupPanelOptions.ShowPopupPanel;
            this.cbePopupPanelLocation.Properties.Items.AddEnum(typeof(TokenEditPopupPanelLocation));
            this.cbePopupPanelLocation.EditValue = this.properties.PopupPanelOptions.Location;
            this.cbeDropDownMode.Properties.Items.AddEnum(typeof(TokenEditDropDownShowMode));
            this.cbeDropDownMode.EditValue = this.properties.DropDownShowMode;
            this.cbeEditMode.Properties.Items.AddEnum(typeof(TokenEditMode));
            this.cbeEditMode.EditValue = this.properties.EditMode;
            this.cbeFontSize.Properties.Items.AddEnum(typeof(TokenEditFontSize));
            this.cbeFontSize.EditValue = TokenEditFontSize.Default;
            this.ceShowItemRemoveButton.Checked = this.properties.ShowRemoveTokenButtons;
            this.cePopupSizeable.Checked = this.properties.PopupSizeable;
            this.cbeGlyphLocation.Properties.Items.AddEnum(typeof(TokenEditGlyphLocation));
            this.cbeGlyphLocation.EditValue = this.properties.TokenGlyphLocation;
            if(this.cbeEditValueSep.Properties.Items.Count > 0) this.cbeEditValueSep.SelectedIndex = 0;
            this.tokenEdit.EditValue = "January";
        }
        // <ceShowDropDown>
        void OnShowDropDownCheckedChanged(object sender, EventArgs e) {
            CheckEdit checkEdit = (CheckEdit)sender;
            this.properties.ShowDropDown = checkEdit.Checked;
        }
        // </ceShowDropDown>
        // <btnClearEventLog>
        void OnClearEventLogClick(object sender, EventArgs e) {
            this.eventsLog.Text = string.Empty;
        }
        // </btnClearEventLog>
        // <tokenEdit>
        void OnTokenEditBeforeShowPopupPanel(object sender, TokenEditBeforeShowPopupPanelEventArgs e) {
            AddEventEntry("BeforeShowPopupPanel", e.Description);
            this.tokenLabelValue.Text = e.Description;
        }
        void OnTokenEditTokenMouseEnter(object sender, TokenEditTokenMouseEnterEventArgs e) {
            AddEventEntry("TokenMouseEnter", e.Description);
        }
        void OnTokenEditTokenMouseHover(object sender, TokenEditTokenMouseHoverEventArgs e) {
            AddEventEntry("TokenMouseHover", e.Description);
        }
        void OnTokenEditTokenMouseLeave(object sender, TokenEditTokenMouseLeaveEventArgs e) {
            AddEventEntry("TokenMouseLeave", e.Description);
        }
        void OnTokenEditValueChanged(object sender, EventArgs e) {
            TokenEdit tokenEdit = (TokenEdit)sender;
            AddEventEntry("EditValueChanged", tokenEdit.EditValue);
            UpdateEditValue();
        }
        void OnTokenEditTokenClick(object sender, TokenEditTokenClickEventArgs e) {
            TokenEdit tokenEdit = (TokenEdit)sender;
            AddEventEntry("TokenClick", e.Description);
        }
        void OnTokenEditTokenDoubleClick(object sender, TokenEditTokenClickEventArgs e) {
            TokenEdit tokenEdit = (TokenEdit)sender;
            AddEventEntry("TokenDoubleClick", e.Description);
        }
        void OnTokenEditCheckStateChanged(object sender, EventArgs e) {
            TokenEdit tokenEdit = (TokenEdit)sender;
            AddEventEntry("TokenCheckStateChanged", null);
        }
        // </tokenEdit>
        // <seDropDownRowCount>
        void OnDropDownRowCountEditValueChanged(object sender, EventArgs e) {
            SpinEdit spinEdit = (SpinEdit)sender;
            this.properties.DropDownRowCount = (int)spinEdit.Value;
        }
        // </seDropDownRowCount>
        // <cbeCheckMode>
        void OnCheckModeSelectedIndexChanged(object sender, EventArgs e) {
            ComboBoxEdit comboBox = (ComboBoxEdit)sender;
            this.properties.CheckMode = (TokenEditCheckMode)comboBox.EditValue;
        }
        // </cbeCheckMode>
        // <seMaxExpandLines>
        void OnMaxExpandLinesEditValueChanged(object sender, EventArgs e) {
            SpinEdit spinEdit = (SpinEdit)sender;
            this.properties.MaxExpandLines = (int)spinEdit.Value;
        }
        // </seMaxExpandLines>
        // <ceShowPopupPanel>
        void OnShowPopupPanelCheckedChanged(object sender, EventArgs e) {
            CheckEdit checkEdit = (CheckEdit)sender;
            this.properties.PopupPanelOptions.ShowPopupPanel = checkEdit.Checked;
        }
        // </ceShowPopupPanel>
        // <cbePopupPanelLocation>
        void OnPopupPanelLocationSelectedIndexChanged(object sender, EventArgs e) {
            ComboBoxEdit comboBox = (ComboBoxEdit)sender;
            this.properties.PopupPanelOptions.Location = (TokenEditPopupPanelLocation)comboBox.EditValue;
        }
        // </cbePopupPanelLocation>
        // <ceShowItemRemoveButton>
        void OnShowItemRemoveButtonCheckedChanged(object sender, EventArgs e) {
            CheckEdit checkEdit = (CheckEdit)sender;
            this.properties.ShowRemoveTokenButtons = checkEdit.Checked;
        }
        // </ceShowItemRemoveButton>
        // <cePopupSizeable>
        void OnPopupSizeableCheckedChanged(object sender, EventArgs e) {
            CheckEdit checkEdit = (CheckEdit)sender;
            this.properties.PopupSizeable = checkEdit.Checked;
        }
        // </cePopupSizeable>
        // <cbeEditValueSep>
        void OnEditValueSepSelectedIndexChanged(object sender, EventArgs e) {
            ComboBoxEdit comboBox = (ComboBoxEdit)sender;
            this.properties.EditValueSeparatorChar = char.Parse(comboBox.EditValue.ToString());
        }
        // </cbeEditValueSep>
        // <cbeDropDownMode>
        void OnDropDownModeSelectedIndexChanged(object sender, EventArgs e) {
            ComboBoxEdit comboBox = (ComboBoxEdit)sender;
            this.properties.DropDownShowMode = (TokenEditDropDownShowMode)comboBox.EditValue;
        }
        // </cbeDropDownMode>
        void UpdateEditValue() {
            string value = tokenEdit.EditValue as string;
            this.editValueLabel.Text = string.Format("Edit Value: {0}", string.IsNullOrEmpty(value) ? "(empty)" : value);
        }
        // <cbeEditMode>
        void OnEditModeSelectedIndexChanged(object sender, EventArgs e) {
            ComboBoxEdit comboBox = (ComboBoxEdit)sender;
            this.properties.EditMode = (TokenEditMode)comboBox.EditValue;
        }
        // </cbeEditMode>
        // <cbeGlyphLocation>
        void OnGlyphAlignmentMode_SelectedIndexChanged(object sender, EventArgs e) {
            ComboBoxEdit comboBox = (ComboBoxEdit)sender;
            this.properties.TokenGlyphLocation = (TokenEditGlyphLocation)comboBox.EditValue;
        }
        // </cbeGlyphLocation>
        // <cbeFontSize>
        void OnFontSizeSelectedIndexChanged(object sender, EventArgs e) {
            Font defaultFont = AppearanceObject.DefaultFont, newFont = null;
            ComboBoxEdit comboBox = (ComboBoxEdit)sender;
            TokenEditFontSize fontSize = (TokenEditFontSize)comboBox.EditValue;
            switch(fontSize) {
                case TokenEditFontSize.Default:
                    newFont = defaultFont;
                    break;
                case TokenEditFontSize.Medium:
                    newFont = new Font(defaultFont.FontFamily, defaultFont.SizeInPoints + 2.5f);
                    break;
                case TokenEditFontSize.Large:
                    newFont = new Font(defaultFont.FontFamily, defaultFont.SizeInPoints + 3.75f);
                    break;
            }
            this.properties.Appearance.Font = this.properties.AppearanceDropDown.Font = newFont;
        }
        // </cbeFontSize>
        void AddEventEntry(string entry, object obj) {
            string val = string.Empty;
            if(obj != null) {
                val = (obj is TokenEdit) ? typeof(TokenEdit).Name : obj.ToString();
            }
            else {
                val = "(null)";
            }
            this.eventsLog.Text = string.Format("{0}: {1}{2}", entry, val, Environment.NewLine) + eventsLog.Text;
        }
    }

    internal enum TokenEditFontSize { Default, Medium, Large }
}
